contracts/tfi-factory/src/contract.rs

```diff
@@ -34,11 +34,19 @@ pub fn instantiate(
         return Err(ContractError::InvalidCommission(msg.default_commission));
     }
 
+    let contract_info_query = QueryRequest::Wasm(WasmQuery::ContractInfo {
+        contract_addr: env.contract.address.to_string(),
+    });
+    let contract_info = deps
+        .querier
+        .query::<ContractInfoResponse>(&contract_info_query)?;
```

Comment:
Something feels off for me. 
`env.contract.address` returns address of this very contract. Since it's just being initialized, doesn't it means that this `ContractInfoResponse` will be empty?
What will be the value of `admin`, or where does it comes from? (maybe it's some internal cosmwasm feature?)

Reply:
...I think ANY test could clarify this. ;)

Reply:
That's a good point.

This admin is the migrate admin, set during instantiation. It _should_ be already there, in the underlying (wasm) `ContractInfo` structure. Now, given that this is precisely being called _during_ instantiation, perhaps it's not. My idea was that, given that instantiate is being called with an `env` that already has an assigned address, then `ContractInfo` is already assigned. But now I'm thinking, perhaps it's assigned, but not available to queries. Will need to dig deep into wasmd to confirm this, or test it in a local blockchain.

As an aside, please beware that there are multiple `ContractInfo`s:
 - wasm `ContractInfo`, which we are interested in here.
 - `cw2` (cw-plus) `ContractInfo`, which contains versioning information, and would perhaps better be called `ContractVersion`, or in any case `ContractVersionInfo`
 - `Env`'s `contract: ContractInfo`, which sadly, only contains the address.

Had an initial version in which the query was done in the `create_pair` call. That is, after instantiation. But moved to this one, so that I can cache the migrate admin. Perhaps I'll need to return to that version.

Reply:
> perhaps it's assigned, but not yet available to queries

Seems racy. That would be risky to push then, contract might fail to initialize on "some weird query" and hours/days will be spent investigating. :)
I would at least wrap those values in some `Options`, to allow contract running.

Regarding extending `Env`, as you said, it's breaking and a lot more work. How important is this?

Reply:
Just tested it, and in fact you are right. The instantiation fails with "error: Querier system error: No such contract: tgrade1hyja4uyjktpeh0fxzuw2fmjudr85rk2qu98fa6nuh6d4qru9l0ssvh4aml: instantiate wasm contract failed".

Will change it to read the contract info in the create_pair call, and store it in an optional field in `Config`.

Reply:
> How important is this?

It's important for mainnet. Being able to upgrade any of these contracts is important, or can become important in the future.

Reply:
Another option could be to pass a migrate admin for the token pairs in `InstantiateMsg`. That is, tfi-factory and tfi-pair migrate admins could be different.

But this is nicer. And more amusing.